File: 155_min_stack.ts
class MinStack {
    arr: number[];
    min: number[];
    constructor() {
        this.arr = [];
        this.min = [];
    }

    push(val: number): void {
        this.arr.push(val);
        if (this.min.length === 0) {
            this.min.push(val);
        } else {
            this.min.push(Math.min(val, this.getMin()));
        }
    }

    pop(): void {
        this.arr.pop();
        this.min.pop();
    }

    top(): number {
        return this.arr[this.arr.length - 1];
    }

    getMin(): number {
        return this.min[this.min.length - 1];
    }
}

File: 8thJune2024/155_min_stack.cpp
#include <algorithm>
#include <bits/stdc++.h>
using namespace std;

class MinStack {
  stack<int> s;
  stack<int> min_stack;

public:
  MinStack() {}

  void push(int val) {
    s.push(val);
    if (min_stack.size() == 0) {
      min_stack.push(val);
      return;
    }
    min_stack.push(min(val, min_stack.top()));
  }

  void pop() {
    int temp = s.top();
    s.pop();
    min_stack.pop();
  }

  int top() { return s.top(); }

  int getMin() { return min_stack.top(); }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */
